plugins/recaptcha/class.recaptcha.plugin.php

```diff
@@ -0,0 +1,246 @@
+        echo $sender->fetchView('captcha', 'display', 'plugins/recaptcha');
+    }
+
+    /**
+     * Hook to validate captchas
+     *
+     * @param Gdn_PluginManager $sender
+     * @return boolean
+     * @throws Exception
+     */
+    public function captcha_validate_handler($sender) {
+        $valid = $sender->EventArguments['captchavalid'];
+
+        $recaptchaResponse = Gdn::request()->post('g-recaptcha-response');
+        if (!$recaptchaResponse) {
+            return $valid = false;
+        }
+
+        return $valid = $this->validateCaptcha($recaptchaResponse);
+    }
+
+    /**
+     * Hook to return captcha submission
+     *
+     * @param Gdn_PluginManager $sender
+     */
+    public function captcha_get_handler($sender) {
+        $recaptchaResponse = Gdn::request()->post('g-recaptcha-response');
+        if ($recaptchaResponse) {
+            $sender->EventArguments['captchatext'] = $recaptchaResponse;
+        }
+    }
+
+    /**
+     * Display reCAPTCHA entry field
+     *
+     * THIS METHOD ECHOS DATA
+     *
+     * @param Gdn_Form $sender
+     * @return string
+     */
+    public function gdn_form_captcha_handler($sender) {
+
+        // Tell the form that we've got this
+        $sender->EventArguments['Handled'] = true;
+
+        if (!$this->getPrivateKey() || !$this->getPublicKey()) {
+            echo '<div class="Warning">' . t('reCAPTCHA has not been set up by the site administrator in registration settings. This is required to register.') .  '</div>';
+        }
+
+        // Google whitelist https://developers.google.com/recaptcha/docs/language
+        $whitelist = ['ar', 'bg', 'ca', 'zh-CN', 'zh-TW', 'hr', 'cs', 'da', 'nl', 'en-GB', 'en', 'fil', 'fi', 'fr', 'fr-CA', 'de', 'de-AT', 'de-CH', 'el', 'iw', 'hi', 'hu', 'id', 'it', 'ja', 'ko', 'lv', 'lt', 'no', 'fa', 'pl', 'pt', 'pt-BR', 'pt-PT', 'ro', 'ru', 'sr', 'sk', 'sl', 'es', 'es-419', 'sv', 'th', 'tr', 'uk', 'vi'];
+
+        // Use our current locale against the whitelist.
+        $language = Gdn::locale()->language();
+        if (!in_array($language, $whitelist)) {
+            $language = (in_array(Gdn::locale()->Locale, $whitelist)) ? Gdn::locale()->Locale : false;
+        }
+
+        Gdn::controller()->Head->addScript('https://www.google.com/recaptcha/api.js?hl=' . $language);
+
+        $attributes = array('class' => 'g-recaptcha', 'data-sitekey' => $this->getPublicKey());
+
+        // see https://developers.google.com/recaptcha/docs/display for details
+        $this->EventArguments['Attributes'] = &$attributes;
+        $this->fireEvent('BeforeCaptcha');
+
+        echo '<div '. attribute($attributes) . '></div>';
+    }
+
+    /**
+     * Hook plugin enabled
+     *
+     * Move old keys to new location.
+     */
+    public function setup() {
+        // Move private key
+        if (c('Garden.Registration.CaptchaPrivateKey')) {
+            saveToConfig('Plugins.Recaptcha.PrivateKey', c('Garden.Registration.CaptchaPrivateKey'));
```

Comment:
If we had it all to do over, we'd have never created the `Plugins.` config namespace, so let's not make a new one under it.